include/frozen/string.h

```diff
@@ -70,12 +71,91 @@ template <> struct elsa<string> {
   }
 };
 
+template<std::size_t size>
+class string_needle {
+  bits::cvector<std::ptrdiff_t, size> cache_;
+  std::array<char, size> data_;
+
+  static constexpr
+  bits::cvector<std::ptrdiff_t, size>
+  kmp(std::array<char, size> needle) {
+    char const* Needle = needle.data();
+    std::ptrdiff_t cnd = 0;
+    bits::cvector<std::ptrdiff_t, size> cache;
+    for(std::size_t pos = 0; pos < size; ++pos)
+      cache[pos] = -1;
+
+    for(std::size_t pos = 1; pos < size; ++pos) {
+      if(Needle[pos] == Needle[cnd]) {
+        cache[pos] = cache[cnd];
+        cnd += 1;
+      }
+      else {
+        cache[pos] = cnd;
+        cnd = cache[cnd];
+        while(cnd >= 0 && Needle[pos] != Needle[cnd])
+          cnd = cache[cnd];
+        cnd += 1;
+      }
+    }
+    return cache;
+  }
+
+  public:
+    constexpr string_needle(std::array<char, size> data)
+      : cache_{kmp(data)}, data_(data) {}
+
+    template<std::size_t... I>
+    constexpr string_needle(char const data[size], std::index_sequence<I...>)
+      : string_needle(std::array<char, size>{{data[I]...}})
+    {}
+    constexpr string_needle(char const data[size])
+      : string_needle(data, std::make_index_sequence<size>())
+    {}
+
+    constexpr char operator[](std::size_t i) const {
+      return (&std::get<0>(data_))[i];
+    }
+    constexpr std::ptrdiff_t step(std::size_t i) const {
+      return cache_[i];
+    }
+};
+
+template<std::size_t N>
+std::size_t strstr(std::string const& haystack, string_needle<N> const& needle) {
```

Comment:
Maybe a comment mentioning that this function is using the KMP search algorithm may be useful when we'll reading back this code in several years...

Reply:
Yes! I'll a wikipedia ref and provide an interface similar to the std::search you've been pointing out.